vlog4j-parser/src/test/java/org/semanticweb/vlog4j/syntax/parser/EntityTest.java

```diff
@@ -0,0 +1,190 @@
+	}
+
+	public void abstractConstantToStringRoundTripTest() throws ParsingException {
+		AbstractConstantImpl f = new AbstractConstantImpl("f");
+		AbstractConstantImpl a = new AbstractConstantImpl("1");
+		Predicate p = Expressions.makePredicate("p", 1);
+		Fact f1 = Expressions.makeFact(p, Arrays.asList(f));
+		Fact f2 = Expressions.makeFact(p, Arrays.asList(a));
+		assertEquals(f1, RuleParser.parseFact(f1.toString()));
+		assertEquals(f2, RuleParser.parseFact(f2.toString()));
+	}
+
+	@Test
+	public void ruleToStringRoundTripTest() throws ParsingException {
+		Constant c = Expressions.makeAbstractConstant("c");
+		Variable x = Expressions.makeUniversalVariable("X");
+		Variable y = Expressions.makeUniversalVariable("Y");
+		Variable z = Expressions.makeExistentialVariable("Z");
+		PositiveLiteral atom1 = Expressions.makePositiveLiteral("p", x, c);
+		PositiveLiteral atom2 = Expressions.makePositiveLiteral("p", x, y);
+		PositiveLiteral headAtom1 = Expressions.makePositiveLiteral("q", x, z);
+		Conjunction<Literal> bodyLiterals = Expressions.makeConjunction(atom1, atom2);
+		Conjunction<PositiveLiteral> headPositiveLiterals = Expressions.makePositiveConjunction(headAtom1);
+		Rule rule1 = new RuleImpl(headPositiveLiterals, bodyLiterals);
+		assertEquals(rule1, RuleParser.parseRule(rule1.toString()));
+	}
+
+	@Test
+	public void conjunctionToStringRoundTripTest() throws ParsingException {
+		Constant c = Expressions.makeAbstractConstant("c");
+		Variable x = Expressions.makeUniversalVariable("X");
+		Variable y = Expressions.makeUniversalVariable("Y");
+		Variable z = Expressions.makeExistentialVariable("Z");
+		NegativeLiteral atom1 = Expressions.makeNegativeLiteral("p", x, c);
+		PositiveLiteral atom2 = Expressions.makePositiveLiteral("p", x, y);
+		PositiveLiteral headAtom1 = Expressions.makePositiveLiteral("q", x, z);
+		Conjunction<Literal> bodyLiterals = Expressions.makeConjunction(atom1, atom2);
+		Conjunction<PositiveLiteral> headPositiveLiterals = Expressions.makePositiveConjunction(headAtom1);
+		Rule rule1 = new RuleImpl(headPositiveLiterals, bodyLiterals);
+		assertEquals(rule1, RuleParser.parseRule(rule1.toString()));
+	}
+
+	@Test
+	public void literalToStringRoundTripTest() throws ParsingException {
+		Constant c = Expressions.makeAbstractConstant("c");
+		Variable x = Expressions.makeUniversalVariable("X");
+		Variable z = Expressions.makeExistentialVariable("Z");
+		NegativeLiteral atom1 = Expressions.makeNegativeLiteral("p", x, c);
+		PositiveLiteral headAtom1 = Expressions.makePositiveLiteral("q", x, z);
+		Rule rule1 = Expressions.makeRule(headAtom1, atom1);
+		assertEquals(rule1, RuleParser.parseRule(rule1.toString()));
+	}
+
+	@Test
+	public void datatypeDoubleConstantToStringRoundTripTest() throws ParsingException {
+		String shortDoubleConstant = "12.345E67";
+		assertEquals(shortDoubleConstant,
+				RuleParser.parseFact("p(\"12.345E67\"^^<http://www.w3.org/2001/XMLSchema#double>).").getArguments()
```

Comment:
just reuse the variable `shortDoubleConstant` inside the other strings, otherwise we have to read each string and make sure it is identical to `shortDoubleConstant`